File: 프로그래머스/lv3/43164. 여행경로/여행경로.py
from collections import defaultdict

def dfs(graph, path, visit):
    if path:
        to = path[-1]
        
        if graph[to]:
            path.append(graph[to].pop(0))
        else:
            visit.append(path.pop())
        
        dfs(graph, path, visit)

    return visit[::-1]

def solution(tickets):
    graph = defaultdict(list)
    
    for start, end in tickets:
        graph[start].append(end)
    for key in graph.keys():
        graph[key].sort()

    return dfs(graph, ["ICN"], [])

File: 프로그래머스/3/43164. 여행경로/여행경로.js
function solution(tickets) {
    let routes = {}
    
    tickets.forEach(([start, end]) => {
        if (routes[start]) {
            routes[start].push(end)
            routes[start].sort()
        } else {
            routes[start] = [end]
        }
    })
    
    let visitedRoute = []
    
    function dfs(node) {
        const neighbors = routes[node]

        while (neighbors && neighbors.length > 0) {
            const nextNode = neighbors.shift()
            dfs(nextNode);
        }

        visitedRoute.push(node)
    }
    
    dfs('ICN')
    
    return visitedRoute.reverse()
}